File: example/motors/motors_set_zero.cpp
#include "robot.h"

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <thread>

constexpr float kVelocity = 0.314f;  // rad/s
constexpr float kAcceleration = 3.14f; // rad/s^2

static std::atomic<bool> exit_flag(false);

int main()
{
    constexpr auto kControlPeriod = std::chrono::microseconds(1000);
    std::signal(SIGINT, [](int) { exit_flag.store(true); });

    Robot robot;

    auto next_tick = std::chrono::steady_clock::now();

    // 通过 robot.motors 以受限速度和加速度持续下发零位目标。
    while (!exit_flag.load())
    {
        for (Motor& motor : robot.motors)
        {
            motor.pos_vel_acc(0.0f, kVelocity, kAcceleration);
        }
        robot.send();

        for (Motor& motor : robot.motors)
        {
            const motor_state_t* state = motor.get_motor_state();
            if (state != nullptr)
            {
                std::printf("ID: %2d, mode: %2d, fluat: %2d, pos: %2.3f, vel: %2.3f, tor: %2.3f\n",
                            motor.get_id(), state->mode, state->fault,
                            state->position, state->velocity, state->torque);
            }
        }
        next_tick += kControlPeriod;
        std::this_thread::sleep_until(next_tick);
    }

    robot.stop();
    robot.send();
    return 0;
}

File: example/python/motors/motors_set_zero.py
import time

from hightorque_robot import Robot


def main():
    kControlPeriod = 0.001
    robot = Robot("../robot_param/robot_config.yaml")

    robot.motor_zero_pos_reset()

    next_tick = time.monotonic()
    while True:
        robot.request_motor_state()

        for motor in robot.motors:
            state = motor.get_motor_state()
            print(
                "ID: {:2d}, mode: {:2d}, fault: {:2d}, pos: {:2.3f}, "
                "vel: {:2.3f}, tor: {:2.3f}".format(
                    motor.get_id(),
                    state.mode,
                    state.fault,
                    state.position,
                    state.velocity,
                    state.torque,
                )
            )

        next_tick += kControlPeriod
        time.sleep(max(0.0, next_tick - time.monotonic()))

    robot.stop()
    robot.send()


if __name__ == "__main__":
    main()
